esi_leap/api/controllers/v1/utils.py

```diff
@@ -0,0 +1,130 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_policy import policy as oslo_policy
+from oslo_utils import uuidutils
+
+from esi_leap.common import exception
+from esi_leap.common import policy
+from esi_leap.common import statuses
+from esi_leap.objects import contract
+from esi_leap.objects import offer
+from esi_leap.resource_objects import resource_object_factory as ro_factory
+
+
+def get_authorized_offer(uuid_or_name, status_filter, cdict):
+    if uuidutils.is_uuid_like(uuid_or_name):
+        o = offer.Offer.get(uuid_or_name)
+        offer_objs = []
+
+        if (status_filter and o.status == status_filter) or \
+                status_filter is None:
+            try:
+                policy.authorize('esi_leap:offer:offer_admin', cdict, cdict)
+                offer_objs.append(o)
+            except oslo_policy.PolicyNotAuthorized:
+                pass
+
+    else:
+        try:
+            policy.authorize('esi_leap:offer:offer_admin', cdict, cdict)
+            offer_objs = offer.Offer.get_all({'name': uuid_or_name,
+                                              'status': status_filter})
+
+        except oslo_policy.PolicyNotAuthorized:
+            offer_objs = offer.Offer.get_all({'name': uuid_or_name,
+                                              'project_id': cdict['project_id'],
+                                              'status': status_filter})
+
+    if len(offer_objs) == 0:
+        raise exception.OfferNotFound(offer_uuid=uuid_or_name)
+    elif len(offer_objs) > 1:
+        raise exception.OfferDuplicateName(name=uuid_or_name)
+
+    verify_resource_permission(cdict, offer_objs[0].to_dict())
```

Comment:
I don't think this belongs here, right? It's only called within the offer post method?

Reply:
Its also called in the offer deletion

Reply:
Does an offer deletion need to verify resource permission?

Reply:
The reason it was originally there was to ensure that an admin can delete the offer as well. But we changed things a bit back so I'm not quite sure now. Is it possible for a resource to ever change hands after offer creation?

Reply:
That's technically possible, but I don't think it's a case we need to cover; it'll be awkward to have an offer in some sort of limbo state caught between owner1 and owner2.